log/file.go

```diff
@@ -0,0 +1,145 @@
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+// defaultLogfilePath is a default logfile path.
+const defaultLogfilePath = "/tmp/k6.log"
+
+// filepathRE matches a valid filepath.
+var filepathRE = regexp.MustCompile(`^(.*)\/([^\/]+)$`)
+
+// fileHook is a hook to handle writing to local files.
+type fileHook struct {
+	lock *sync.Mutex
+
+	path   string
+	levels []logrus.Level
+}
+
+func getDefaultFileHook() *fileHook {
+	return &fileHook{
+		lock:   new(sync.Mutex),
+		path:   defaultLogfilePath,
+		levels: logrus.AllLevels,
+	}
+}
+
+// FileHookFromConfigLine returns new fileHook hook.
+func FileHookFromConfigLine(line string) (logrus.Hook, error) {
+	hook := getDefaultFileHook()
+
+	if line != "file" {
+		parts := strings.SplitN(line, "=", 2)
+		if parts[0] != "file" {
+			return nil, fmt.Errorf("logfile configuration should be in the form `file=path-to-local-file` but is `%s`", line)
+		}
+
+		if err := hook.parseArgs(line); err != nil {
+			return nil, err
+		}
+	}
+
+	return hook, nil
+}
+
+func (h *fileHook) parseArgs(line string) error {
+	tokens, err := tokenize(line)
+	if err != nil {
+		return fmt.Errorf("error while parsing logfile configuration %w", err)
+	}
+
+	for _, token := range tokens {
+		key := token.key
+		value := token.value
+
+		switch key {
+		case "file":
+			if !filepathRE.MatchString(value) {
+				return fmt.Errorf("path to logfile is invalid: %s", value)
+			}
+
+			h.path = value
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown logfile config key %s", key)
+		}
+	}
+
+	return nil
+}
+
+// Fire writes the log file to defined path.
+func (h *fileHook) Fire(entry *logrus.Entry) error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	return h.write(entry)
+}
+
+// write writes a log line to a logfile.
+func (h *fileHook) write(entry *logrus.Entry) (err error) {
+	dir := filepath.Dir(h.path)
+	if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to create dir for logfile: %s %w", h.path, err)
+	}
+
+	file, err := os.OpenFile(h.path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0o600)
+	if err != nil {
+		return fmt.Errorf("failed to open logfile: %s %w", h.path, err)
+	}
```

Comment:
Doing in this way will be slow. `FileHookFromConfigLine` should create and open the file once, it should accept a context and listen to `Done` for closing the file. The lines should be flushed asynchronously mostly as the `loki` hook is already doing.

Reply:
I completely agree with you. I'll refactor it.

Reply:
I made it more asynchronous. Was trying to keep it as simple as possible.